File: lib/constructs/security/cloudtrail-construct.ts
import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as cloudtrail from 'aws-cdk-lib/aws-cloudtrail';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as kms from 'aws-cdk-lib/aws-kms';
import { SecurityConfig } from '../../config/security-config';

/**
 * CloudTrail Construct for comprehensive API call logging
 * Records all API calls made to AWS accounts for security auditing
 */
export class CloudTrailConstruct extends Construct {
  public readonly trail: cloudtrail.Trail;
  public readonly logBucket: s3.Bucket;
  public readonly logGroup: logs.LogGroup;

  constructor(scope: Construct, id: string, encryptionKey: kms.Key) {
    super(scope, id);

    // S3 bucket for CloudTrail logs
    this.logBucket = new s3.Bucket(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-CloudTrail-Bucket`,
      {
        bucketName: `${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-cloudtrail-logs-${this.node.addr.slice(-8)}`,
        encryption: s3.BucketEncryption.KMS,
        encryptionKey: encryptionKey,
        versioned: true,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        lifecycleRules: [
          {
            id: 'DeleteOldCloudTrailLogs',
            expiration: cdk.Duration.days(SecurityConfig.LOG_RETENTION_DAYS),
          },
        ],
      }
    );

    // CloudWatch Log Group for CloudTrail
    this.logGroup = new logs.LogGroup(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-CloudTrail-Logs`,
      {
        logGroupName: `/aws/cloudtrail/${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}`,
        retention: logs.RetentionDays.ONE_YEAR,
        encryptionKey: encryptionKey,
      }
    );

    // CloudTrail with comprehensive logging
    this.trail = new cloudtrail.Trail(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-CloudTrail`,
      {
        trailName: `${SecurityConfig.RESOURCE_PREFIX}-CloudTrail`,
        bucket: this.logBucket,
        cloudWatchLogGroup: this.logGroup,
        includeGlobalServiceEvents: true,
        isMultiRegionTrail: true,
        enableFileValidation: true,
        encryptionKey: encryptionKey,
        s3KeyPrefix: 'cloudtrail-logs/',
      }
    );

    // Add advanced event selectors for comprehensive logging
    const cfnTrail = this.trail.node.defaultChild as cloudtrail.CfnTrail;
    cfnTrail.addPropertyOverride('AdvancedEventSelectors', [
      {
        Name: 'All Management Events',
        FieldSelectors: [
          {
            Field: 'eventCategory',
            Equals: ['Management'],
          },
        ],
      },
      {
        Name: 'S3 Data Events',
        FieldSelectors: [
          {
            Field: 'eventCategory',
            Equals: ['Data'],
          },
          {
            Field: 'resources.type',
            Equals: ['AWS::S3::Object'],
          },
        ],
      },
      {
        Name: 'Lambda Data Events',
        FieldSelectors: [
          {
            Field: 'eventCategory',
            Equals: ['Data'],
          },
          {
            Field: 'resources.type',
            Equals: ['AWS::Lambda::Function'],
          },
        ],
      },
    ]);
  }
}

File: archive/cdk-java/Pr2382/lib/src/main/java/app/constructs/CloudTrailConstruct.java
package app.constructs;

import software.amazon.awscdk.services.cloudtrail.Trail;
import software.amazon.awscdk.services.cloudtrail.ReadWriteType;
import software.amazon.awscdk.services.s3.IBucket;
import software.amazon.awscdk.services.kms.IKey;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.logs.RetentionDays;
import software.constructs.Construct;
import java.util.UUID;
import app.config.EnvironmentConfig;

/**
 * CloudTrail construct that implements comprehensive logging and monitoring.
 * This construct creates CloudTrail with encryption, log file validation,
 * and CloudWatch integration for real-time monitoring.
 */
public class CloudTrailConstruct extends Construct {
    
    private final Trail cloudTrail;
    private final LogGroup logGroup;
    
    public CloudTrailConstruct(final Construct scope, final String id, 
                              final IBucket s3Bucket, final IKey kmsKey) {
        super(scope, id);
        
        // Create CloudWatch log group for CloudTrail
        this.logGroup = createLogGroup(kmsKey);
        
        // Create CloudTrail with comprehensive logging
        this.cloudTrail = createCloudTrail(s3Bucket, kmsKey);
    }
    
    /**
     * Creates a CloudWatch log group for CloudTrail logs with encryption.
     */
    private LogGroup createLogGroup(final IKey kmsKey) {
    // Use a short random suffix on the physical log group name to avoid
    // collisions with existing global log group names in other accounts/environments.
    final String randomSuffix = UUID.randomUUID().toString().substring(0, 8);
    final String physicalName = EnvironmentConfig.getResourceName("cloudtrail", "log-group") + "-" + randomSuffix;

    return LogGroup.Builder.create(this, EnvironmentConfig.getResourceName("cloudtrail", "log-group"))
        .logGroupName(physicalName)
        .retention(RetentionDays.ONE_YEAR) // Retain logs for compliance
        .encryptionKey(kmsKey)
        .build();
    }
    
    /**
     * Creates CloudTrail with comprehensive security and monitoring settings.
     */
    private Trail createCloudTrail(final IBucket s3Bucket, final IKey kmsKey) {
        // Create a basic multi-region CloudTrail with file validation and CloudWatch integration.
        return Trail.Builder.create(this, EnvironmentConfig.getResourceName("cloudtrail", "trail"))
                .trailName(EnvironmentConfig.getResourceName("cloudtrail", "trail"))
                .bucket(s3Bucket)
                .s3KeyPrefix("cloudtrail-logs/")
                .cloudWatchLogGroup(logGroup)
                .enableFileValidation(true)
                .includeGlobalServiceEvents(true)
                .isMultiRegionTrail(true)
                .managementEvents(ReadWriteType.ALL)
                .build();
    }
    
    // Getters
    public Trail getCloudTrail() { return cloudTrail; }
    public LogGroup getLogGroup() { return logGroup; }
}